patches/tModLoader/Terraria/Item.TML.cs

```diff
@@ -170,6 +172,68 @@ public static int NewItem(IEntitySource source, Rectangle rectangle, int Type, i
 		public static int NewItem(IEntitySource source, Vector2 position, int Type, int Stack = 1, bool noBroadcast = false, int prefixGiven = 0, bool noGrabDelay = false, bool reverseLookup = false)
 			=> NewItem(source, (int)position.X, (int)position.Y, 0, 0, Type, Stack, noBroadcast, prefixGiven, noGrabDelay, reverseLookup);
 
+		/// <summary>
+		/// <inheritdoc cref="Item.NewItem(IEntitySource, int, int, int, int, int, int, bool, int, bool, bool)"/>
+		/// <br/><br/>This particular overload uses an Item instead of just the item type. All modded data will be preserved.
+		/// </summary>
+		public static int NewItem(IEntitySource source, int X, int Y, int Width, int Height, Item item, bool noBroadcast = false, bool noGrabDelay = false, bool reverseLookup = false) {
+			int Type = item.type;
+
+			if (WorldGen.gen)
+				return 0;
+
+			if (Main.rand == null)
+				Main.rand = new UnifiedRandom();
+
+			if (NPCLoader.blockLoot.Contains(Type))
+				return Main.maxItems;
+
+			if (Type > 0 && cachedItemSpawnsByType[Type] != -1) {
+				cachedItemSpawnsByType[Type] += item.stack;
+				return 400;
+			}
```

Comment:
cached spawns should probably not work with the `Item` variant, only the `type` variant.